### PR TITLE
USHIFT-716: add configurable KAS advertise address

### PR DESCRIPTION
**Pull request by pacevedom**
**Which issue(s) this PR addresses**:

Closes #<Issue Number>

### CODE REVIEW

File: docs/howto_config.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ node:
 apiServer:
   subjectAltNames:
     - ""
+  advertiseAddress: ""
```
Comment (pacevedom) [inline docs/howto_config.md:27]: This is exposed so as to enable CNCF certification tests where we need more than one node. Users shouldnt use it, as it defaults to the api server service IP internally.

Comment (dhellmann) [reply docs/howto_config.md:27]: If we put this in the documentation, someone is going to use it. Do we need to include it in this document?

File: packaging/microshift/config.yaml [hunk 2]
```diff
@@ -26,6 +26,9 @@ apiServer:
   # The Subject Alternative Names for the external certificates in API server (defaults to hostname -A)
   #subjectAltNames: []
 
+  # The Kube API Server advertise address for pods to connect to it (defaults to 10.43.0.1)
+  #advertiseAddress: ''
```
Comment (dhellmann) [inline packaging/microshift/config.yaml:30]: Can we remove it from here, too?

File: pkg/cmd/init.go [hunk 3]
```diff
@@ -70,6 +70,15 @@ func certSetup(cfg *config.MicroshiftConfig) (*certchains.CertificateChains, err
 		return nil, err
 	}
 
+	externalCertNames := []string{
+		cfg.NodeName,
+		"api." + cfg.BaseDomain,
+	}
+	externalCertNames = append(externalCertNames, cfg.SubjectAltNames...)
+	if cfg.KASAdvertiseAddress != cfg.NodeIP {
```
Comment (dhellmann) [inline pkg/cmd/init.go:78]: A comment here explaining why we only include the NodeIP if it does not match the kube API server address will help maintainers who are looking at this code later.

File: pkg/cmd/run.go [hunk 4]
```diff
@@ -36,6 +36,7 @@ func addRunFlags(cmd *cobra.Command, cfg *config.MicroshiftConfig) {
 	flags.String("service-cidr", cfg.Cluster.ServiceCIDR, "The IP range in CIDR notation for services in the cluster.")
 	flags.String("service-node-port-range", cfg.Cluster.ServiceNodePortRange, "The port range to reserve for services with NodePort visibility. This must not overlap with the ephemeral port range on nodes.")
 	flags.String("base-domain", cfg.BaseDomain, "The base domain for this cluster.")
+	flags.String("kas-advertise-address", cfg.KASAdvertiseAddress, "The KAS advertise address for pods to connect.")
```
Comment (dhellmann) [inline pkg/cmd/run.go:39]: Do we need a command line flag?

File: pkg/config/config.go [hunk 5]
```diff
@@ -59,11 +59,13 @@ type IngressConfig struct {
...
-	SubjectAltNames []string      `json:"subjectAltNames"`
-	NodeName        string        `json:"nodeName"`
-	NodeIP          string        `json:"nodeIP"`
-	BaseDomain      string        `json:"baseDomain"`
-	Cluster         ClusterConfig `json:"cluster"`
+	SubjectAltNames     []string      `json:"subjectAltNames"`
+	NodeName            string        `json:"nodeName"`
+	NodeIP              string        `json:"nodeIP"`
+	KASAdvertiseAddress string        `json:"kasAdvertiseAddress"`
+	SkipKASInterface    bool          `json:"-"`
```
Comment (dhellmann) [inline pkg/config/config.go:66]: What is SkipKASInterface?

At some point we should come back and add comments to all of these fields. You don't have to do all of them in this PR, but let's at least add good docs for the new things.

File: pkg/controllers/kube-apiserver.go [hunk 6]
```diff
@@ -298,6 +304,20 @@ func (s *KubeAPIServer) Run(ctx context.Context, ready chan<- struct{}, stopped
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
+	if !s.skipLoInterface {
```
Comment (dhellmann) [inline pkg/controllers/kube-apiserver.go:307]: It's not clear where to draw a line for the single responsibility principle here. Should this IP management be done inside the KubeAPIServer service or can it be its own thing that KubeAPIServer depends on?

**Comment by dhellmann** [comment]
Is there a way to add a test case to verify that this solves the problem and so we do not regress with future changes?

File: pkg/controllers/kube-apiserver.go (hunk 6 above)
Comment (pacevedom) [reply pkg/controllers/kube-apiserver.go]: Turned this into a new service. Makes more sense, thanks for the suggestion.

File: docs/howto_config.md (hunk 1 above)
Comment (dhellmann) [reply docs/howto_config.md:27]: That's a good question. @jerpeter1 has been looking at the certification rules, maybe he came across something that would help answer it?

File: pkg/cmd/run.go (hunk 4 above)
Comment (dhellmann) [reply pkg/cmd/run.go:39]: I would like to standardize on just using the configuration file, and dropping the environment variables and command line flags. We can always add this later if we find a reason not to do that.

Comment (pacevedom) [reply pkg/cmd/run.go]: Done

**Comment by pacevedom** [comment]
/override ci/prow/e2e-openshift-conformance-sig-scheduling ci/prow/e2e-openshift-conformance-sig-instrumentation ci/prow/e2e-openshift-conformance-sig-api-machinery

**Pull request merged**